reltoken.php

```diff
@@ -25,7 +34,56 @@ function reltoken_civicrm_tokens(&$tokens) {
       if (strpos($token, '{contact') !== FALSE) {
         $tokenBase = preg_replace('/^\{contact\.(\w+)\}$/', '$1', $token);
         // Must be in the form: $tokens['X']["X.whatever"] where X is not "contact".
-        $tokens['related']["related.{$tokenBase}___reltype_{$hash}"] = "Related ({$relationshipTypeDetails['directionLabel']})::{$label}";
+        $e->entity('related')->register("{$tokenBase}___reltype_{$hash}", "Related ({$relationshipTypeDetails['directionLabel']})::{$label}");
+      }
+    }
+  }
+}
+
+function reltoken_evaluate_tokens(\Civi\Token\Event\TokenValueEvent $e) {
+  $tokens = $e->getTokenProcessor()->getMessageTokens();
+  $contactIDs = $e->getTokenProcessor()->getContextValues('contactId');
+  if (!empty($tokens['related'])) {
+    foreach ($tokens['related'] as $token) {
+      if (strpos($token, '___reltype_')) {
+        $relatedContactIDsPerContact = _reltoken_get_related_contact_ids_per_contact($contactIDs, $token);
+        $relatedContactIDs = array_unique(array_values($relatedContactIDsPerContact));
+        // If you're using a token for a relationship this person doesn't have, just skip it
+        // Otherwise you create a query that crushes the system.
+        if (!$relatedContactIDs) {
+          continue;
+        }
+        $baseToken = preg_replace('/^(.+)___.+$/', '$1', $token);
+        $schema[] = 'contactId';
```

Comment:
Any reason why we don't use `$schema = ['contactId']` here? (i.e., why append to the array repeatedly on each `foreach` iteration instead of defining it once as a single-member array?)